Fixes inconsistent behaviour in portforward

m1kola opened the pull request:
**What type of PR is this?**

/kind flake

**What this PR does / why we need it**:

Fixes address parsing in port-forward (`client-go` & `kubectl`).

The rules for address parsing are:

* Explicitly specified addresses must bind successfully
* `localhost` is pinned to `127.0.0.1` and `::1` and at least one of those must bind successfully

This change also makes output of the command consistent between runs with the same arguments. Previously the command was using the range via map of addresses which sometimes was producing different output because the order of values is not guaranteed in Go.

**Which issue(s) this PR fixes**:

#71017

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```


---

/sig cli
/priority P2

m1kola committed:
Fixes address parsing in port-forward

The rules for address parsing are:

* Explicitly specified addresses must bind successfully
* `localhost` is pinned to `127.0.0.1` and `::1` and at least one of those must bind successfully

This change also makes output of the command consistent
between runs with the same arguments.
Previously the command was using the range via map of addresses
which sometimes was producing different output because the order
of values is not guaranteed in Go.

m1kola commented:
Hm. I just realised that flaky e2e tests do not rely on the order in which "Forwarding from ..." messages appear (they are being parsed with a regexp).

But I still think that this needs to be merged, because it makes the order of "Forwarding from ..." messages consistent between runs.

fedebongio commented:
/assign @roycaihw

roycaihw commented on staging/src/k8s.io/client-go/tools/portforward/portforward.go:
```diff
@@ -117,28 +117,35 @@ type listenAddress struct {
+			parsed[ipv4.address] = ipv4
+			parsed[ipv6.address] = ipv6
+			// IPv4 address for localhost must go first for backward compatibility
+			keyOrder = append(keyOrder, ipv4.address, ipv6.address)
```
nit: you could add two lines of
```
keyOrder = append(keyOrder, ip.address)
```
instead of changing existing lines. I'm not sure which way is more readable.

roycaihw commented on staging/src/k8s.io/client-go/tools/portforward/portforward.go:
```diff
@@ -117,28 +117,35 @@ type listenAddress struct {
-			parsed[address] = listenAddress{address: address, protocol: "tcp4", failureMode: "any"}
+			ip := listenAddress{address: address, protocol: "tcp4", failureMode: "any"}
+			parsed[ip.address] = ip
+			keyOrder = append(keyOrder, ip.address)
```
nit: similar here and below. You could do
```
keyOrder = append(keyOrder, address)
```

m1kola replied:
I changed it. For some reason in diff github shows it in a different place after rebase.

roycaihw commented on test/e2e/kubectl/portforward.go:
```diff
@@ -49,7 +49,7 @@ const (
 // TODO support other ports besides 80
 var (
-	portForwardRegexp = regexp.MustCompile("Forwarding from (127.0.0.1|\\[::1\\]):([0-9]+) -> 80")
+	portForwardRegexp = regexp.MustCompile("Forwarding from 127.0.0.1:([0-9]+) -> 80")
```
Why do we remove ipv6 localhost from matching?

m1kola replied:
I was looking into flaky tests originally and thought that we are not going to need this after my fix, but then I realised that I was wrong (I thought it has something to do with the order of output).

I removed it.

roycaihw reviewed (commented):
The idea of keeping the order same as the input SGTM. An alternative is to sort the slice like the [unit test](https://github.com/kubernetes/kubernetes/blob/3bcbc5da79978fd48525af82da03434f81e1cdd9/staging/src/k8s.io/client-go/tools/portforward/portforward_test.go#L142-L143) does. I guess keeping the order makes more sense.

m1kola commented:
Thanks for the review. I addressed the feedback.

> An alternative is to sort the slice like the [unit test](https://github.com/kubernetes/kubernetes/blob/3bcbc5da79978fd48525af82da03434f81e1cdd9/staging/src/k8s.io/client-go/tools/portforward/portforward_test.go#L142-L143) does. I guess keeping the order makes more sense.

I think, the original author was going to keep the order from input (from `addressesToParse `), but just didn't take into account that `range` over a `map` doesn't guarantee consistent ordering.

I removed sorting in the unit test so it checks order as well now. Sorting there was the reason why we didn't have flakes in unit tests.

roycaihw commented on staging/src/k8s.io/client-go/tools/portforward/portforward_test.go:
```diff
@@ -83,6 +82,17 @@ func TestParsePortsAndNew(t *testing.T) {
 				{protocol: "tcp6", address: "::1", failureMode: "all"},
```
I wonder if it's a bug or if we don't care. 

For this testcase, when address `"localhost"` was processed, the "output" addresses should have been
```
{protocol: "tcp4", address: "127.0.0.1", failureMode: "all"},
{protocol: "tcp6", address: "::1", failureMode: "all"},
```
However the consequent address `"127.0.0.1"` gets processed and overrides the `failureMode` of 127.0.0.1 to `"any"`.

I can imagine the reverse case can also happen. It's orthogonal to this PR so I won't hold it

m1kola replied:
@rowleyajI this is related to [this conversation](https://github.com/kubernetes/kubernetes/pull/46517/files#r179165644) in the original PR about how to handle failures and it seems to be expected behaviour:

* if someone specifies the addresses explicitly - then command should fail if any of the address fail (covered by this test case)
* If no addresses specified explicitly (`localhost`), then allow to silently fail one of them

roycaihw replied:
Right. I'm not convinced that the current implementation is the final goal based on the conversation. But for the current implementation, imagine the two scenarios: 

1. input addresses: `"localhost", "127.0.0.1", "::1"`; output addresses would be 
```
{protocol: "tcp4", address: "127.0.0.1", failureMode: "any"},
{protocol: "tcp6", address: "::1", failureMode: "any"},
```
2. input addresses: `"127.0.0.1", "::1", "localhost"`; output addresses would be 
```
{protocol: "tcp4", address: "127.0.0.1", failureMode: "all"},
{protocol: "tcp6", address: "::1", failureMode: "all"},
```

I'm thinking open an issue to continue the discussion in https://github.com/kubernetes/kubernetes/pull/46517/files#r179165644

m1kola replied:
Yeah, you are right. The special case for `localhost` at the end overrides previously set values.

Please, cc me on the issue, once created.

m1kola commented:
> Should we try to get this in 1.13, since it fixes a flaky test?

@seans3 I think, it will not fix it. See https://github.com/kubernetes/kubernetes/pull/70931#issuecomment-437724893. Sorry, for the confusion.

soltysh reviewed (changes requested):
/hold

liggitt commented on staging/src/k8s.io/client-go/tools/portforward/portforward.go:
```diff
@@ -117,28 +117,35 @@ type listenAddress struct {
 		if address == "localhost" {
 			ip := listenAddress{address: "127.0.0.1", protocol: "tcp4", failureMode: "all"}
 			parsed[ip.address] = ip
+			// IPv4 address for localhost must go first in keyOrder for backward compatibility
```
the issue raised in https://github.com/kubernetes/kubernetes/issues/71017 is legitimate. I think it might be as simple as ensuring we never switch an existing "any" rule to an "all" rule

m1kola replied:
> I don't recall this being a backwards compatibility requirement.

Sorry, for the confusion again. Initially I was looking into flaky tests and though older tests (skew) expect `127.0.0.1` to go first in logs, but then I realised that tests check log output with regexp and that it's not relevant.

I can remove this comment to avoid further confusion.

soltysh replied:
@m1kola yes please remove the comment and add a test verifying the issue mentioned is fixed too.

nikopen commented:
/kind bug
/priority critical-urgent

/cc @soltysh 
comment to be addressed

AishSundar commented:
@m1kola @soltysh @liggit how critical is this PR for 1.13?  we are in code freeze and if possible, I would like to defer this until 1.14. let me know otherwise

/remove-priority critical-urgent

AishSundar commented:
/remove-priority critical-urgent

nikopen commented:
@m1kola though this is almost ready, you were away the last week and it's getting a bit late in the cycle.

moving it to patch fixes and 1.14

/milestone v1.14

m1kola commented:
@nikopen sorry, I didn't reply earlier. I think this PR is ready now. There is a separate issue (#71017). I'll try to explain it below.

@soltysh I removed the comment as discussed [here](https://github.com/kubernetes/kubernetes/pull/70931#discussion_r235444728) to avoid confusion. #71017 is a separate issue.

This PR fixes inconsistent behaviour. Prior changes in #46517 `kubectl port-forward` was producing the following output constantly:

```
Forwarding from 127.0.0.1:8000 -> 80
Forwarding from [::1]:8000 -> 80
```

After the change, if you run the command multiple times with the same arguments you can sometimes get the output as above, but sometimes you can get the output in a different order: 

```
Forwarding from [::1]:8000 -> 80
Forwarding from 127.0.0.1:8000 -> 80
```

This PR only makes the output of the command to be reproducible between runs with the same arguments.

#71017 is about how `kubectl port-forward` handles errors when it can't bind to an address. See this thread https://github.com/kubernetes/kubernetes/pull/70931#discussion_r233277999 and links in it.

goblain commented on staging/src/k8s.io/client-go/tools/portforward/portforward.go:
```diff
@@ -112,34 +112,37 @@ func parsePorts(ports []string) ([]ForwardedPort, error) {
-			parsed[address] = listenAddress{address: address, protocol: "tcp6", failureMode: "any"}
-		} else {
-			return nil, fmt.Errorf("%s is not a valid IP", address)
+		ips, err := net.LookupIP(address)
```
unfortunately LookupIP is not a reliable way of getting loopback address(es). It relies on name resolution, which in turn will use ie. /etc/hosts so even if you have both `127.0.0.1` and `::1` but only one of them in hosts file, this method will report only the IP defined in hosts.

https://github.com/kubernetes/kubernetes/pull/69578/files#diff-43e7f6a6df5c4f7000f26222d9afb0cfR399 is the solution suggested by @lavalamp for this.

m1kola replied:
Thanks for looking into this.

I have some concerns about this approach in context of `kubectl`, but l I'm not an expert in this, so, please, excuse me if I say something stupid :)

In the code example above, the function gets the loopback interface by hardcoded name `lo`. How will this work on different operating systems? On Linux machines I see `lo` interface, but on my Mac OS I see `lo0` interface. Will it work on Windows?

I think we can loop over all interfaces and check if an interface is has both `FlagUp` and `FlagLoopback` flags to avoid hardcoding the name, but this raises more questions:

1. What if we find more than one loopback interface?
2. Should we try to bind to all IPs from all interfaces we find?
3. If users explicitly runs `--address localhost`, do resolve using DNS or do we look for IPs on loopback interfaces?

goblain replied:
the implementation in original code is for `localhost` to simply translate to `127.0.0.1` and/or `::1`. I don't think we should support any other addresses for non-explicit, default `localhost`. If you want to bind to some IP on unusual loopback, you can just provide the IP in the `--address` flag. Maybe just search if each of these two localhost addresses exist on some `FlagUp` interface.

Btw. LookupIP has another side effect of reintroducing ability to use names in `--address` values, while it was explicitly removed in original PR as was seen as expanding the scope of kubectl / client responsibility too much (like using `--address some.domain.tld`).

Wonder if it wouldn't be the best to return an error like `can't bind twice to the same address` if the IP shows up twice (ie. from `127.0.0.1,localhost`) instead of updating and messing up failure mode in this edge case.

m1kola replied:
> Maybe just search if each of these two localhost addresses exist on some `FlagUp` interface.

Makes perfect sense. I can do that.

> Wonder if it wouldn't be the best to return an error like `can't bind twice to the same address` if the IP shows up twice (ie. from `127.0.0.1,localhost`) instead of updating and messing up failure mode in this edge case.

I like the idea of returning the "can't bind twice" error if for some reason the same address appears twice. Let's be explicit.

---

Before I go further, do I understand correctly that we still want to allow users something like `--address localhost,10.19.21.23` (it's from `kubectl port-forward --help` message)? There were suggestions in https://github.com/kubernetes/kubernetes/pull/46517#discussion_r223808638 and https://github.com/kubernetes/kubernetes/pull/46517#discussion_r179140223 from @lavalamp and @liggitt about not allowing passing `localhost` into `--address` (if I interpreted their comments correctly).

Dropping this will make the implementation much clearer, but I'm afraid that v1.13.0-rc.1 has already been released and it's too late to drop it.

liggitt replied:
> the implementation in original code is for localhost to simply translate to 127.0.0.1 and/or ::1. I don't think we should support any other addresses for implicit, default localhost. If you want to bind to some IP on unusual loopback, you can just provide the IP in the --address flag.

I agree with that, and the behavior should be documented in the flag help.

>> Maybe just search if each of these two localhost addresses exist on some FlagUp interface.
>
> Makes perfect sense. I can do that.

I'd actually try to reduce the amount of interface inspection done, not increase it. I think the following behavior would be reasonable:

1. explicitly specified addresses must bind successfully
2. `localhost` is pinned to `127.0.0.1` and `::1` and at least one of those must bind successfully

2 can be accomplished by changing the localhost address parsing to never override an explicitly specified address:

  ```go
  if address == "localhost" {
    if _, exists := parsed["127.0.0.1"]; !exists {
      ip := listenAddress{address: "127.0.0.1", protocol: "tcp4", failureMode: "all"}
      parsed[ip.address] = ip
    }
    if _, exists := parsed["::1"]; !exists {
      ip = listenAddress{address: "::1", protocol: "tcp6", failureMode: "all"}
      parsed[ip.address] = ip
    }
  } ...
  ```

m1kola commented:
@soltysh I updated the PR. Please, take another look.

soltysh reviewed (approved):
/lgtm
/approve
@liggitt any comments?

liggitt commented on staging/src/k8s.io/client-go/tools/portforward/portforward.go:
```diff
@@ -117,28 +117,38 @@ type listenAddress struct {
+
+	parsedCount := len(parsed)
+	addresses := make([]listenAddress, parsedCount)
+	for id, mapKey := range keyOrder[:parsedCount] {
```
this will truncate incorrectly with args like this:

localhost,127.0.0.1,10.10.10.1

parsedCount will be 3, and keyOrder will have "127.0.0.1" present twice

if you drop the keyOrder logic and revert changes from line 137 on, this will do what you want

m1kola replied:
@liggitt oh, right... Do I understand correctly that we do not worry about order of the output (it will be different from time to time)?

I did add this `keyOrder` logic to always have the same output for the same input. It was previously causing issues in end-to-end tests, but now these tests do not rely on order. Just wondering if any of the users rely on it or not. I can't think of any useful use cases, so I'm happy to drop it.

liggitt replied:
> Do I understand correctly that we do not worry about order of the output (it will be different from time to time)?

if you want to sort the addresses by IP before returning to get a stable order, that's fine

m1kola replied:
Oh, yes. That makes more sense than trying to save the order specified in the input (in `addressesToParse`).

Thanks for the feedback. I'll update the PR on the weekend.

liggitt commented:
/hold

m1kola commented:
@soltysh @liggitt I updated the PR. Please, take another look.

liggitt commented:
/hold cancel
/lgtm
/approve
